poetry/packages/project_package.py

```diff
@@ -8,6 +8,23 @@
 
 class ProjectPackage(Package):
     def __init__(self, name, version, pretty_version=None):
+        if isinstance(version, str) and version.startswith('attr:'):
+            version_pkg, _, version_attr = (
+                version[6:].strip().rpartition('.')
+            )
+            if not version_pkg:
+                version_pkg = 'builtins'
+            version_getter = getattr(
+                __import__(version_pkg, fromlist=(version_pkg, )),
+                version_attr,
+            )
+            version = version_getter()
```

Comment:
People may want to import a variable instead of a function.  The setuptools_scm documentation shows an example in the programmatic usage section https://github.com/pypa/setuptools_scm#programmatic-usage
Checking if the attribute is callable would allow for this. https://docs.python.org/3/library/functions.html#callable

Reply:
That's different. Also the callable may as well return that constant.

Reply:
Constant declaration is already supported by setting it in `pyproject.toml`. There should be only one way to do it.

Reply:
The variable is dynamically determined; I do not think this is already supported by `pyproject.toml`.  I think adding a small function wrapper that returns a variable is an acceptable thing.  But a small if check for whether or not it is `callable` would support both.  Tiny quality of life thing that saves writing a tiny function for people who copy/paste from setuptools_scm docs.

Reply:
You're suggesting to put import-time evaluation in place, which is exactly what doomed setuptools. And now because of backwards compat they cannot fix this.
That's why I'm `-1` here.
If one wants to access the installed version info from their code we could provide them a helper function to do so but let's not do import-time version detection. After all, app needs the version retrieved from pkg_resources in installed mode, it shouldn't invoke version-guessing flows in this case.

Reply:
What is the difference in what is accomplished between a variable that holds the version, and a function that evaluates to a variable that holds the version?  I was suggesting this:
```python
if callable(version_getter):
    version = version_getter()
else:
    version = version_getter
```

Reply:
The difference is that it encourages users to have 

```python
version = blah_blah()
```

called on during the module import, which is bad.